### PR TITLE
[SPARK-7675][ML][PYSpark] sparkml params type conversion

### PR DESCRIPTION
**Pull request by holdenk**
From JIRA:
Currently, PySpark wrappers for spark.ml Scala classes are brittle when accepting Param types. E.g., Normalizer's "p" param cannot be set to "2" (an integer); it must be set to "2.0" (a float). Fixing this is not trivial since there does not appear to be a natural place to insert the conversion before Python wrappers call Java's Params setter method.

A possible fix will be to include a method "_checkType" to PySpark's Param class which checks the type, prints an error if needed, and converts types when relevant (e.g., int to float, or scipy matrix to array). The Java wrapper method which copies params to Scala can call this method when available.

This fix instead checks the types at set time since I think failing sooner is better, but I can switch it around to check at copy time if that would be better. So far this only converts int to float and other conversions (like scipymatrix to array) are left for the future.

### CODE REVIEW

**Comment by holdenk** [comment]
ping @jkbradley

**Comment by holdenk** [comment]
cc @yanboliang & @mengxr since you also seem to do a lot of work in the param code.

**Comment by holdenk** [comment]
also @gweidner

**Comment by holdenk** [comment]
cc @davies & @yu-iskw

**Comment by holdenk** [comment]
ping

**Comment by holdenk** [comment]
cc @jkbradley

**Comment by holdenk** [comment]
cc @davies

File: python/pyspark/ml/param/shared.py [hunk 1]
```diff
@@ -603,19 +603,19 @@ class HasSolver(Params):
...
-        self.solver = Param(self, "solver", "the solver algorithm for optimization. If this is not set or empty, default value is 'auto'.")
+        self.solver = Param(self, "solver", "the solver algorithm for optimization. If this is not set or empty, default value is 'auto'.", str)
         self._setDefault(solver='auto')
 
     def setSolver(self, value):
         """
         Sets the value of :py:attr:`solver`.
         """
-        self._paramMap[self.solver] = value
+        self._set(**{"solver": value})
```
Comment (davies) [inline python/pyspark/ml/param/shared.py]: Why not this one?
```
self._set(solver=value)
```

Comment (holdenk) [inline python/pyspark/ml/param/shared.py]: Sure that makes sense - I'll change the codegen script.

File: python/pyspark/ml/param/__init__.py [hunk 2]
```diff
@@ -247,7 +248,27 @@ def _set(self, **kwargs):
         Sets user-supplied params.
         """
         for param, value in kwargs.items():
-            self._paramMap[getattr(self, param)] = value
+            p = getattr(self, param)
+            if (p.expectedType is not None):
+                if p.expectedType is None or type(value) == p.expectedType or value is None:
```
Comment (sethah) [inline python/pyspark/ml/param/__init__.py]: I believe the `p.expectedType is None` check in the second if statement is frivolous. Couldn't this be condensed into a single if else like:

```python
if p.expectedType is not None and type(value) != p.expectedType and value is not None:
    try:
        ...
else:
    self._paramMap[getattr(self, param)] = value
```

Comment (holdenk) [inline python/pyspark/ml/param/__init__.py]: Thats a good point, I'll simplify the branching.

**Comment by holdenk** [comment]
I've fixed the _set call ( @davies ) and simplified the branching ( @sethah ).

File: python/pyspark/ml/param/_shared_params_code_gen.py [hunk 3]
```diff
@@ -105,44 +109,45 @@ def get$Name(self):
...
     print("from pyspark.ml.param import Param, Params\n\n")
     shared = [
-        ("maxIter", "max number of iterations (>= 0).", None),
-        ("regParam", "regularization parameter (>= 0).", None),
-        ("featuresCol", "features column name.", "'features'"),
-        ("labelCol", "label column name.", "'label'"),
-        ("predictionCol", "prediction column name.", "'prediction'"),
+        ("maxIter", "max number of iterations (>= 0).", None, int),
+        ("regParam", "regularization parameter (>= 0).", None, float),
+        ("featuresCol", "features column name.", "'features'", None),
```
Comment (jkbradley) [inline python/pyspark/ml/param/_shared_params_code_gen.py]: str

File: python/pyspark/ml/param/_shared_params_code_gen.py [hunk 4]
```diff
@@ -105,44 +109,45 @@ def get$Name(self):
...
-        ("inputCol", "input column name.", None),
-        ("inputCols", "input column names.", None),
-        ("outputCol", "output column name.", "self.uid + '__output'"),
-        ("numFeatures", "number of features.", None),
+         "should be treated as confidences, not precise probabilities.", "'probability'", str),
+        ("rawPredictionCol", "raw prediction (a.k.a. confidence) column name.", "'rawPrediction'",
+         str),
+        ("inputCol", "input column name.", None, str),
+        ("inputCols", "input column names.", None, None),
+        ("outputCol", "output column name.", "self.uid + '__output'", None),
```
Comment (jkbradley) [inline python/pyspark/ml/param/_shared_params_code_gen.py]: str

File: python/pyspark/ml/param/_shared_params_code_gen.py [hunk 5]
```diff
@@ -105,44 +109,45 @@ def get$Name(self):
...
-        ("tol", "the convergence tolerance for iterative algorithms.", None),
-        ("stepSize", "Step size to be used for each iteration of optimization.", None),
+         "E.g. 10 means that the cache will get checkpointed every 10 iterations.", None, int),
+        ("seed", "random seed.", "hash(type(self).__name__)", int),
+        ("tol", "the convergence tolerance for iterative algorithms.", None, float),
+        ("stepSize", "Step size to be used for each iteration of optimization.", None, float),
         ("handleInvalid", "how to handle invalid entries. Options are skip (which will filter " +
          "out rows with bad values), or error (which will throw an errror). More options may be " +
-         "added later.", None),
+         "added later.", None, bool),
```
Comment (jkbradley) [inline python/pyspark/ml/param/_shared_params_code_gen.py]: str

File: python/pyspark/ml/param/__init__.py [hunk 6]
```diff
@@ -247,7 +248,24 @@ def _set(self, **kwargs):
...
+                try:
+                    # Try and do "safe" conversions that don't lose information
+                    if p.expectedType == float:
+                        self._paramMap[getattr(self, param)] = float(value)
+                    # Python 3 unified long & int
+                    elif p.expectedType == int and type(value).__name__ == 'long':
+                        self._paramMap[getattr(self, param)] = value
+                    else:
+                        raise Exception(
+                            "Provided type {0} incompatable with type {1} for param {2}"
```
Comment (jkbradley) [inline python/pyspark/ml/param/__init__.py]: typo: "compatible"

**Comment by jkbradley** [comment]
@holdenk I like this approach.  Just made 1 pass.

**Comment by holdenk** [comment]
@jkbradley thanks! I've made the suggested changes :)

**Comment by jkbradley** [comment]
@holdenk Thanks!  Just one more item I forgot: could you please add a unit test for this in ml/tests.py?

**Comment by holdenk** [comment]
Sure - I figured it was already pretty well covered by the existing tests which use the params but I'll add some explicit unit tests for this.

**Comment by jkbradley** [comment]
It'd be nice to have a test or two which failed before your PR but work now.  They don't need to touch all (or any) algorithms; you should hopefully be able to use a fake Params subclass which uses one of the float Params, etc.

**Comment by holdenk** [comment]
I've added a test using LogisitcRegression - I figured it would be better to use a real model so as to validate that it got picked up OK on the Java side too.

**Comment by holdenk** [comment]
jenkins, retest this please.

**Comment by jkbradley** [comment]
LGTM.  Thanks for the PR!
Merging with master

**Comment by sethah** [comment]
I am thinking about how we can avoid getting `java.lang.ClassCastException` when passing incorrect `Param` types through to Scala. Is there any reason we cannot make `expectedType` a required argument? Then we set it if the type equals expected type, and try to convert it if not. If it cannot be converted then we raise an exception. It would be nice to avoid users getting somewhat cryptic, non-intuitive py4j errors. Thoughts?

**Comment by holdenk** [comment]
I mean right now we don't have anything for lists so that would be a blocking thing before we could make it required.

**Pull request closed without merging**